Major refactor to Java launcher library

TheKodeToad opened the pull request:
solonovamax helped a lot with this. Thank you!

Changes:
- Rename enums to SHOUTING_SNAKE_CASE.
- Add more exit codes.
- Switch back from java.util.Logger to a custom one, this time with better code using an enum. The reason for this is the launcher did not colourise the logs like with the old behaviour. While JUL is better in terms of OOP, it doesn't have all the necessary log levels and in the end would require more code to make the log messages look nice. The API is also not the best. Using logging frameworks would just create bloat for no reason and probably create issues.
- Split the `onesix` launcher into `standard` and `legacy` (Standard, Legacy and AbstractLauncher were introduced).
- Reintroduce MultiMC's printing the chosen launcher, but from the C++ code instead.
- Reformat the code using Eclipse. It has a good formatter, okay.
- Updated license headers.
- Split Utils into different classes.
- Use MethodHandle instead of reflection where possible and practical.
- Fix warnings instead of ignoring them. If you use the correct JDK none will be created (for the source code - there is one unrelated one).
- Removed LauncherFactory again. Sorry icelimetea, but I personally think a factory is not needed for two launcher types.
- Remove constructor in NewLaunch. Why have a field when you can have a local variable? I generally also like to avoid instances of the main class.
- Remove feeding the classpath through stdin to the Java launcher. It's already passed into the java executable with the -cp flag, and NewLaunch completely ignores it.

TheKodeToad committed:
Various tweaks to the Java component of the launcher

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

flowln reviewed (commented):
this creates a few new warnings:

```bash
org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
        mcParams = params.allSafe("param", Collections.EMPTY_LIST);
                                 ^
  required: String,List<String>
  found: String,List
org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
        mcParams = params.allSafe("param", Collections.EMPTY_LIST);
                                                      ^
  required: List<String>
  found:    List
org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
        mcParams = params.allSafe("param", Collections.EMPTY_LIST);
                                 ^
  required: List<String>
  found:    List
org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
        traits = params.allSafe("traits", Collections.EMPTY_LIST);
                               ^
  required: String,List<String>
  found: String,List
org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
        traits = params.allSafe("traits", Collections.EMPTY_LIST);
                                                     ^
  required: List<String>
  found:    List
org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
        traits = params.allSafe("traits", Collections.EMPTY_LIST);
                               ^
  required: List<String>
  found:    List
```

they should probably be fixed, even if it incurs in some performance overhead. this isn't exactly a hot path anyways :3

TheKodeToad commented:
> this creates a few new warnings:
> 
> ```shell
> org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
>         mcParams = params.allSafe("param", Collections.EMPTY_LIST);
>                                  ^
>   required: String,List<String>
>   found: String,List
> org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
>         mcParams = params.allSafe("param", Collections.EMPTY_LIST);
>                                                       ^
>   required: List<String>
>   found:    List
> org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
>         mcParams = params.allSafe("param", Collections.EMPTY_LIST);
>                                  ^
>   required: List<String>
>   found:    List
> org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
>         traits = params.allSafe("traits", Collections.EMPTY_LIST);
>                                ^
>   required: String,List<String>
>   found: String,List
> org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
>         traits = params.allSafe("traits", Collections.EMPTY_LIST);
>                                                      ^
>   required: List<String>
>   found:    List
> org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
>         traits = params.allSafe("traits", Collections.EMPTY_LIST);
>                                ^
>   required: List<String>
>   found:    List
> ```
> 
> they should probably be fixed, even if it incurs in some performance overhead. this isn't exactly a hot path anyways :3

I have fixed this now. Even if the performance impact is nothing to lose sleep about (well, it is for me), it's still generally preferred to use emptyList when you want an immutable empty list. Fortunately I managed to keep using this method in a safer way.

Anyway, does anyone have any idea of what I should change with the license headers, if anything?

txtsd commented:
`java.applet.Applet` is deprecated in newer Java versions. Is this required for older modloaders? And can it be replaced with whatever the modern alternative for applets is?

TheKodeToad commented:
> `java.applet.Applet` is deprecated in newer Java versions. Is this required for older modloaders? And can it be replaced with whatever the modern alternative for applets is?

Old versions of Minecraft used Java applets so they could be run in web browsers. I think there is a standalone main class though. However, since there are other compatibility issues with newer versions of Java, I can't see many reasons to replace it.

solonovamax commented:
Perhaps it might be better to split the applet/legacy launcher and the default launcher into separate classes?

There [already exists support for different launcher types](https://github.com/PrismLauncher/PrismLauncher/blob/develop/libraries/launcher/org/prismlauncher/LauncherFactory.java#L50-L68), so why not use it?

That way you can `@Suppress` all instances of deprecated for the entire class, while only affecting the applet code.

solonovamax commented:
Also, in the line of some improvements to the java code, why not use the `ParseException` class here instead of `IllegalStateException`? https://github.com/PrismLauncher/PrismLauncher/blob/deb42d1b1e71c29f7bf385c349d9c7772e09b243/libraries/launcher/org/prismlauncher/impl/OneSixLauncher.java#L93

TheKodeToad commented:
Okay, I will try to improve the code more - maybe more of a refactor. I was trying to avoid it because other projects (it was mainly one though) didn't like it when I kept adding to my pull requests.

Time to fire up an IDE that isn't totally terrible ;)

txtsd commented:
> Okay, I will try to improve the code more - maybe more of a refactor. I was trying to avoid it because other projects (it was mainly one though) didn't like it when I kept adding to my pull requests.
> 
> Time to fire up an IDE that isn't totally terrible ;)

Take your time and make your changes. We won't merge it until it's deemed ready.

solonovamax commented:
> Okay, I will try to improve the code more - maybe more of a refactor. I was trying to avoid it because other projects (it was mainly one though) didn't like it when I kept adding to my pull requests.

I was gonna make my own pr with a bunch of code cleanup changes later as well, bc I noticed quite a few things that could be changed

TheKodeToad commented:
I think I've done enough for this pull request. I've implemented https://github.com/PrismLauncher/PrismLauncher/pull/291#issuecomment-1291395239, as well as replaced some string splitting with substringing (just personal preference, but I think it's best to avoid regex when searching for single characters - and it can have side-effects if the character appears more times than you want). Maybe I'll open one aimed at improving the legacy launcher.

Also, somebody's been here before.

![image](https://user-images.githubusercontent.com/57493648/198069133-413c2bb4-d671-425f-bef2-f18d675b2e39.png)

I reverted the change only for one of the fields which is changed.

No idea why re-requesting a review removed txtsd's.

solonovamax commented:
few things I just noticed:

1. The license header
   ```
   /*
    *  PolyMC - Minecraft Launcher
    *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
    *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
    *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
   ```
   should be changed for
   ```
   /*
    *  Prism Launcher - Minecraft Launcher
    *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
    *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
    *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
   ```
2. You removed `OneSixLauncher.java` from the launcher code, however, I'm pretty sure that the "onesix launcher" is referenced in many other places in the codebase. May want to change that to just reference the "launcher". (Search for "onesix" or "one six" in the codebase)


Also, I'd very much like to contribute some of my own changes to this, but I want to commit them myself, so they're under my name (As they're slightly more significant than the previous changes I suggested). Would that be possible? If so, how do you want me to do that? I could make a fork of your fork, then PR to it, so that way it can be merged into this branch.

TheKodeToad commented:
> few things I just noticed:
> 
>     1. The license header
>        ```
>        /*
>         *  PolyMC - Minecraft Launcher
>         *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
>         *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
>         *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        should be changed for
>        ```
>        /*
>         *  Prism Launcher - Minecraft Launcher
>         *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
>         *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
>         *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
>        ```
> 
>     2. You removed `OneSixLauncher.java` from the launcher code, however, I'm pretty sure that the "onesix launcher" is referenced in many other places in the codebase. May want to change that to just reference the "launcher". (Search for "onesix" or "one six" in the codebase)
> 
> 
> Also, I'd very much like to contribute some of my own changes to this, but I want to commit them myself, so they're under my name (As they're slightly more significant than the previous changes I suggested). Would that be possible? If so, how do you want me to do that? I could make a fork of your fork, then PR to it, so that way it can be merged into this branch.

Oh dear, this is getting confusing. I'll give you push access. I'm not sure if people normally do that, but since you're not part of the Prism Launcher organisation it would be that or a fork fork (and this repo is already a fork fork).

solonovamax commented:
:+1: alright

~~I pinky promise not to `git push --force`~~

Scrumplex commented:
> emptyList when you want an immutable empty list

We used to use emptyList() for everything but some lists can be modified conditionally so it should actually all be mutable.

EDIT: nvm you already saw my commit :D

solonovamax committed:
Get rid of singleton, and refactor LauncherFactory to be a static class

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

solonovamax committed:
Code refactors

- Refactor LauncherFactory.LauncherProvider to LauncherFactory
- Refactor all launcher related components to launcher package
- some basic code cleanup
- Rename all, allSafe -> getList and first, firstSafe -> getString
- Rename Utils -> LegacyUtils

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

solonovamax committed:
Cleanup EntryPoint code

- Don't return an int from listen(). An enum is preferred.
- Make parseLine() static, and pass Parameters to it.

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

solonovamax committed:
Code cleanup to net.minecraft.Launcher

No code logic has been changed, only:
- add @Override annotatons
- change setVisible(boolean b) -> setVisible(boolean visible)
- Change block commend on class -> javadoc comment

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

TheKodeToad committed:
Fix formatting

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Attempt to sort out copyright headers

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Better variable naming

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Remove the need for LegacyUtils

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Excessive safety

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Create launcher.impl.legacy

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

solonovamax commented on libraries/launcher/net/minecraft/Launcher.java:
```diff
@@ -84,7 +84,6 @@ public Launcher(Applet applet) {
     }
 
     public Launcher(Applet applet, URL documentBase) {
-        super();
```
We should explicitly have `super();` in the constructor. It's good practice, as otherwise there's just an implied super call. (Plus, my IDE complains when it's not there)

TheKodeToad replied:
In the majority of code styles, this isn't an issue. I don't see any reason to add something that doesn't change anything, since it's already fairly apparent that the class extends something else. In my experience, IDEs don't warn about it by default. However, I'm not really that bothered either way.

solonovamax commented:
@TheKodeToad I was going to keep `LegacyUtils`, because I wanted to move a bit of the messy logic happening in the `LegacyLauncher` into that, to clean it up.

You alright if I revert that commit later?

solonovamax commented:
Also, avoid force pushes if possible, because it forces me to have to rebase/re-clone my entire branch

TheKodeToad commented:
> @TheKodeToad I was going to keep `LegacyUtils`, because I wanted to move a bit of the messy logic happening in the `LegacyLauncher` into that, to clean it up.
> 
> You alright if I revert that commit later?

It seems weird to couple a class with a utils one. I feel like it would be more logical to move some of the messy logic into other methods.

> Also, avoid force pushes if possible, because it forces me to have to rebase/re-clone my entire branch

Oops, sorry. I had to because I forgot signing off. I think I've got the hang of it now.

Edit: OMG I keep pushing and then realising my mistake. Still, it shouldn't make a difference unless you already have the commits on your local copy - rebasing is normally required even without force pushing.

TheKodeToad committed:
Make main static

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

Inline local variable

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad commented:
Only now do I remember that method handles are not available in Java 7.

solonovamax commented:
> Edit: OMG I keep pushing and then realising my mistake. Still, it shouldn't make a difference unless you already have the commits on your local copy - rebasing is normally required even without force pushing.

If you want to make sure to sign every single commit, add the following to your `.gitconfig`:
```
[user]
    name = [YOUR NAME HERE]
    email = [YOUR EMAIL HERE]
    signingkey = [YOUR SIGNING KEY HERE]
[commit]
    gpgsign = true
```

Then, it will sign every single one of your commits by default (so you don't need to specify `-S` on the `git commit` command.)

TheKodeToad commented:
> > Edit: OMG I keep pushing and then realising my mistake. Still, it shouldn't make a difference unless you already have the commits on your local copy - rebasing is normally required even without force pushing.
> 
> If you want to make sure to sign every single commit, add the following to your `.gitconfig`:
> 
> ```
> [user]
>     name = [YOUR NAME HERE]
>     email = [YOUR EMAIL HERE]
>     signingkey = [YOUR SIGNING KEY HERE]
> [commit]
>     gpgsign = true
> ```
> 
> Then, it will sign every single one of your commits by default (so you don't need to specify `-S` on the `git commit` command.)

This is for signing, not signing off.

Anyway, I just realised not to panic because turns out MethodHandles were in Java 7 and I am stupid!

solonovamax committed:
Cleanup splitting string into a pair

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

TheKodeToad commented:
What is the point of using an array? Then it's pretty similar to just using String::split. I chose to use substring instead due to it being clearer in my opinion.

solonovamax commented:
@TheKodeToad it's like 2x-3x faster lol

I did some profiling, and on my laptop `String#split` gets around 3k-3.5k ops/ms, whereas the `String#indexOf` based split gets ~8.5k ops/ms

I want to try and optimize that startup loop, as in my experience I've noticed prism takes a bit to launch mc once you press "launch". Though, idk if that's the java code or the cpp code (haven't profiled it)

I also want to make that code as minimal as possible. I'm probably going to remove the hash map in `LauncherFactory`, and provide some optimizations to the `BufferedReader` used in reading the params (possibly use an `InputStream` directly, instead)

Also, I would have used a `Pair` or a `Tuple`, except java doesn't have either of those in the jdk.

TheKodeToad commented:
Well, creating an array has a tiny performance overhead too, but it sounds like a pretty good trade-off considering it's already faster than before.

Most of the startup time is probably from the C++ side, but I'd be interested how long it takes to get from org.prismlauncher...EntryPoint to net.minecraft...Main.

BufferedReaders afaik are pretty much the best thing to use when reading line-by-line.

solonovamax commented:
> BufferedReaders afaik are pretty much the best thing to use when reading line-by-line.

iirc I did some profiling a while ago, and using `InputStream` directly can actually vastly outperform it

solonovamax committed:
Get rid of gross HashMap being used by `LauncherFactory` and instead use *fancy* enum

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

solonovamax committed:
Cleanup launcher classes

Cleanup a bunch of the code in launcher classes
- Migrate the majority of the reflection to ReflectionUtils
- Decrease logic in AbstractLauncher
- Add logging to launcher classes at FINE level
- make mcParams in AbstractLauncher an immutable list to prevent runtime manipulation
  - StandardLauncher instead copies the list to modify it

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

solonovamax committed:
More informative exceptions in entrypoint

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

TheKodeToad commented:
I've heard that Java enums have plenty of performance issues. There's likely some *background HashMap* that we never hear about. I feel like a C-style enum would make sense - it could be passed from the C++ code with ease.

This is how I'd do it. But the integer would have to be parsed which isn't great.

```c++
enum
{
    STANDARD, LEGACY
}

launchScript += "launcher ";

// use legacy launcher if the traits are set
if (profile->getTraits().contains("legacyLaunch") || profile->getTraits().contains("alphaLaunch"))
    launchScript += LEGACY;
else
    launchScript += STANDARD;

launchScript += "\n";
```


```java
public static final int STANDARD = 0, LEGACY = 1;

public static Launcher getLauncher(int id) {
    switch (id) {
        case STANDARD:
            return new StandardLauncher();
        case LEGACY:
            return new LegacyLauncher();
    }
}
```

If we used a raw input stream, though, you can just do in.read() - '0' to get the enum ID. This would work for 10 launchers. *Maybe this is going too far though*.

solonovamax committed:
Fix license headers

- Update license headers in several files to remove multimc apache reference, when unneeded
  - LauncherFactory: we've entirely rewritten this class at this point, so it's fully under GPL code
  - Launcher: this interface contains zero logic, and only a single method named `launch`, so I doubt you can actually copyright that
  - LauncherProvider: same as above
  - ParseException, ParameterNotFoundException: same as above; this class contains almost no logic (And the logic that was added is under GPL)
  - ReflectionUtils, StringUtils: add license header
  - Everything else: modify the program name in the license header from "PolyMC - Minecraft Launcher" to "Prism Launcher"

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

TheKodeToad commented:
I did not have to suppress this many warnings. Maybe you could just tweak your IDE settings on this project? Many of these are not standard warnings and just create more warnings on most compilers which don't understand them.

TheKodeToad committed:
Try to improve consistency
- Makes code formatting more consistent with the C++ codebase. Probably removes some trailing whitespace. Maybe it would be best to commit an Eclipse or IntelliJ code format preferences file?
- Removes obscure suppressions. I personally think it's better to only suppress warnings that javac complains about. Suppressing a lot of non-standardised warnings (many of them turned off by default even in IntelliJ) just creates needless clutter.
- Fixes some trivial warnings instead of suppressing them. serialVersionUID is sort of stupid, but I'd rather mentally ignore it or just fix it if it's really that annoying.

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Less needless verbosity, and return fallback
Needlessly verbose commit message ik.

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Fix WrongMethodTypeException. Invoke exact don't work like that

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Attempt to mimic clang-format

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Sorry for reverting a lot but copying a list just hurts me

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
So turns out we can have nice things

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
*opinionated change* Remove LauncherFactory
Factories in OOP are a classic example of over-enginneering. When you only have two launchers I personally think that it's not very useful.

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Remove arguments being passed twice
Passing the classpath into stdin has no effect. Java is already provided the classpath with -cp, which pretty much takes up the largest part of the arguments anyway, which leads me to wonder, what's the point of stdin arguments at all?

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Just ignore empty lines

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
More standard code formatting profile

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Fix the warnings properly

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
More consistent logging text

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

solonovamax commented on libraries/launcher/org/prismlauncher/utils/ReflectionUtils.java:
```diff
@@ -48,7 +48,6 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
-@SuppressWarnings("removal")
```
Why was this removed?

`Applet` is a deprecated api, and the compiler will produce warnings otherwise.

solonovamax commented on libraries/launcher/org/prismlauncher/launcher/impl/legacy/LegacyFrame.java:
```diff
@@ -77,10 +77,10 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
-@SuppressWarnings({ "serial", "removal" })
```
Why was this removed? `Applet` is a deprecated api, so it produces compiler warnings

solonovamax commented on libraries/launcher/org/prismlauncher/EntryPoint.java:
```diff
@@ -116,7 +116,6 @@ private static ExitCode listen() {
             String line;
 
             while (preLaunchAction == PreLaunchAction.PROCEED) {
-                // noinspection NestedAssignment
```
Why are you removing the inspection suppression comments?

solonovamax reviewed (changes requested):
Compiler warnings & suppression inspection changes

solonovamax commented:
If you're compiling with jdk 8 or later, it will produce warnings. However, it is configured to output jdk 7 compatible class files even if you build with jdk 8, so we should keep the warning suppressions.

TheKodeToad commented:
The warnings go away if you are using Java 8 or lower. Java 9 was the version that deprecated these classes. Suppressing these warnings was just causing more warnings for me.

I thought I'd already said, but I removed the inspection comments as that's something you can tweak in your IDE settings and these do not affect `javac` which is what is actually being used for compilation.

TheKodeToad committed:
Try to use more standard exit codes

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Implicit is generally prefered

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Remove last case of printStackTrace

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
More license fixing

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Add StringUtils to CMakeLists.txt

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Always use this for consistency

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

Scrumplex commented:
Changing PolyMC to Prism Launcher is fine, as that is not part of the copyright header.
Copyright does not care about projects, it cares about documents and authors.

TheKodeToad commented:
> Changing PolyMC to Prism Launcher is fine, as that is not part of the copyright header. Copyright does not care about projects, it cares about documents and authors.

It's inconsistent with the C++ source files though.

solonovamax commented:
> It's inconsistent with the C++ source files though.

Then the cpp source files should be changed, but that's outside the scope of this PR as we're currently focusing on the java launcher component

TheKodeToad committed:
Revert switch to JUL for better colours

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

Scrumplex commented:
> It's inconsistent with the C++ source files though.

Well. We are slowly changing it to `Prism Launcher` in all files we modify. We won't do it at once for all files, but rather just for stuff we actually modify.

TheKodeToad commented:
I have managed to squash and drop a lot of commits. It should be ready for review now. The code is at a stage where to see something wrong with it you'd have to be not me.

TheKodeToad committed:
A lot more cleanup

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

Scrumplex commented on libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java:
```diff
@@ -33,48 +33,67 @@
+/**
+ * Used to create a print stream that redirects to Log.
+ */
+final class LogPrintStream extends PrintStream {
```
I don't like this. Don't we have another way of doing this?

TheKodeToad replied:
There are alternatives, but they would likely miss some of the nuances of the stock implementation - which is fairly complex.
I can understand why this class isn't ideal, though.

Scrumplex replied:
From what I can tell this is only used for stacktraces. I think it would make everyone's life easier if stacktraces don't have any prefix anyway.

TheKodeToad replied:
The prefixes are used to signal the text colour to the log viewer.
The log viewer *does* automatically convert stacktraces to red though.
Just means the fatal stacktraces won't get the black background.

Scrumplex replied:
IMO it's fine. A fatal stacktrace generally means that the game/launcher will quit immediately. So it should be fine

Scrumplex commented on libraries/launcher/org/prismlauncher/utils/StringUtils.java:
```diff
@@ -0,0 +1,49 @@
+        return new String[] { str.substring(0, splitPoint), str.substring(splitPoint + 1) };
+    }
+
+}
```
drop this file

Scrumplex commented on libraries/launcher/org/prismlauncher/EntryPoint.java:
```diff
@@ -52,113 +54,126 @@
-        } catch (Exception e) {
-            LOGGER.log(Level.SEVERE, "Exception caught from launcher.", e);
+            default:
+                String[] pair = StringUtils.splitStringPair(' ', input);
```
```suggestion
                String[] pair = input.split(" ", 2);
```

TheKodeToad replied:
This is just reverting the change that was made for performance. However, if it was changed to " " instead of "\\s+" that probably would still improve performance (since I think single-character patterns are optimised).

Let me know what you think.

Scrumplex replied:
Yeah, but I don't really see why we would need to have our own function here, as we should keep the code concise. I think we have been using split for the last few years without anyone caring, so I don't think the performance gain is worth the extra code.

TheKodeToad committed:
Make requested changes and utilise AssertionError

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Remove dirty printStackTrace hacks

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

Scrumplex reviewed (approved):
Thanks a lot!
